be/src/io/cache/fs_file_cache_storage.cpp

```diff
@@ -326,91 +442,139 @@ Status FSFileCacheStorage::upgrade_cache_dir_if_necessary() const {
+                        std::string key_prefix =
+                                Path(_cache_base_path) / cache_key.substr(0, KEY_PREFIX_LENGTH);
+                        bool exists = false;
+                        auto exists_status = fs->exists(key_prefix, &exists);
+                        if (!exists_status.ok()) {
+                            LOG(WARNING) << "Failed to check directory existence: " << key_prefix
+                                         << ", error: " << exists_status.to_string();
+                            ++failure_count;
+                            continue;
+                        }
+                        if (!exists) {
+                            auto create_status = fs->create_directory(key_prefix);
+                            if (!create_status.ok() &&
+                                create_status.code() != TStatusCode::type::ALREADY_EXIST) {
+                                LOG(WARNING) << "Failed to create directory: " << key_prefix
+                                             << ", error: " << create_status.to_string();
+                                ++failure_count;
+                                continue;
+                            }
+                        }
+                        auto rename_status = fs->rename(key_it->path(), key_prefix / cache_key);
+                        if (rename_status.ok() ||
+                            rename_status.code() == TStatusCode::type::DIRECTORY_NOT_EMPTY) {
+                            ++rename_count;
+                        } else {
+                            LOG(WARNING)
+                                    << "Failed to rename directory from " << key_it->path().native()
+                                    << " to " << (key_prefix / cache_key).native()
+                                    << ", error: " << rename_status.to_string();
+                            ++failure_count;
+                            continue;
                         }
-                    }
-                    auto rename_status = fs->rename(key_it->path(), key_prefix / cache_key);
-                    if (rename_status.ok()) {
-                        ++rename_count;
-                    } else {
-                        LOG(WARNING)
-                                << "Failed to rename directory from " << key_it->path().native()
-                                << " to " << (key_prefix / cache_key).native()
-                                << ", error: " << rename_status.to_string();
-                        return rename_status;
                     }
                 }
+            } catch (std::filesystem::filesystem_error& e) {
+                LOG(WARNING) << "Error occurred while upgrading file cache directory: "
+                             << key_it->path() << " err: " << e.what();
+                ++failure_count;
             }
         }
 
         auto rebuild_dir = [&](std::filesystem::directory_iterator& upgrade_key_it) -> Status {
```

Comment:
this procedure is unncecssary
we can finish renaming and appending _0 simutaneously.